src/main/java/ir/sahab/elasticsearchrule/ElasticsearchBase.java

```diff
@@ -0,0 +1,135 @@
+package ir.sahab.elasticsearchrule;
+
+import ir.sahab.cleanup.Cleanups;
+import org.elasticsearch.action.admin.cluster.health.ClusterHealthResponse;
+import org.elasticsearch.action.admin.indices.template.put.PutIndexTemplateRequest;
+import org.elasticsearch.action.admin.indices.template.put.PutIndexTemplateResponse;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.client.transport.TransportClient;
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.health.ClusterHealthStatus;
+import org.elasticsearch.common.network.NetworkModule;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.index.reindex.ReindexPlugin;
+import org.elasticsearch.node.InternalSettingsPreparer;
+import org.elasticsearch.node.Node;
+import org.elasticsearch.node.NodeValidationException;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.transport.Netty4Plugin;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.transport.client.PreBuiltTransportClient;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+class ElasticsearchBase {
+
+    private final Lock lock = new ReentrantLock();
+    private final String clusterName;
+    private Path tempDirectory;
+    private TransportClient transportClient;
+    private TransportAddress transportAddress;
+    private Node server;
+
+    ElasticsearchBase(String clusterName) {
+        this.clusterName = clusterName;
+    }
+
+    protected void setupElasticsearchServer()
+            throws IOException, NodeValidationException, ExecutionException, InterruptedException {
+        lock.lock();
```

Comment:
Add comment why we need this locking mechanism.

Reply:
I'm not sure, but I will add that It is usable for concurrent tests.